File: packages/react-native-brownfield/ios/ReactNativeBrownfieldModule.swift
internal import React

@objcMembers
public class ReactNativeBrownfieldModuleImpl: NSObject {
  static public func setPopGestureRecognizerEnabled(_ enabled: Bool) {
    let userInfo = ["enabled": enabled]
    DispatchQueue.main.async {
      NotificationCenter.default.post(name: Notification.Name.togglePopGestureRecognizer, object: nil, userInfo: userInfo)
    }
  }

  static public func popToNative(animated: Bool) {
    let userInfo = ["animated": animated]
    DispatchQueue.main.async {
      NotificationCenter.default.post(name: Notification.Name.popToNative, object: nil, userInfo: userInfo)
    }
  }

  static public func postMessage(_ message: String) {
    let userInfo: [String: Any] = ["message": message]
    DispatchQueue.main.async {
      NotificationCenter.default.post(name: Notification.Name.brownfieldMessageFromJS, object: nil, userInfo: userInfo)
    }
  }
}

File: packages/react-native-brownfield/android/src/oldarch/ReactNativeBrownfieldModule.kt
package com.callstack.reactnativebrownfield

import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class ReactNativeBrownfieldModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  companion object {
    var shouldPopToNative: Boolean = false
    private const val LOG_TAG = "ReactNativeBrownfieldModule"

    fun emitMessageFromNative(text: String) {
      Log.w(
          LOG_TAG,
          "ReactNativeBrownfieldModule::emitMessageFromNative only supports the New Architecture. This call is ineffective and will not cause any messages to be emitted."
      )
    }
  }

  @ReactMethod
  fun popToNative(animated: Boolean) {
    shouldPopToNative = true
    onBackPressed()
  }

  @ReactMethod
  fun setPopGestureRecognizerEnabled(enabled: Boolean) {
    shouldPopToNative = enabled
  }

  @ReactMethod
  fun setHardwareBackButtonEnabled(isFirstRoute: Boolean) {
    shouldPopToNative = isFirstRoute
  }

  @ReactMethod
  fun postMessage(message: String) {
    ReactNativeBrownfield.shared.dispatchMessage(message)
  }

  @ReactMethod
  fun addListener(eventName: String) {}

  @ReactMethod
  fun removeListeners(count: Double) {}

  private fun onBackPressed() {
    reactApplicationContext.currentActivity?.runOnUiThread {
      reactApplicationContext.currentActivity?.onBackPressed()
    }
  }

  override fun getName(): String {
    return "ReactNativeBrownfield"
  }
}
